tests/gstreamer_mqtt/unittest_mqtt_w_helper.cc

```diff
@@ -0,0 +1,886 @@
+  EXPECT_EQ (ret, GST_STATE_CHANGE_SUCCESS);
+  EXPECT_EQ (cur_state, GST_STATE_PLAYING);
+
+  GstMqttTestHelper::getInstance ().setFailDisconnect (FALSE);
+
+  ret = gst_element_set_state (pipeline, GST_STATE_NULL);
+  EXPECT_NE (ret, GST_STATE_CHANGE_FAILURE);
+
+  ret = gst_element_get_state (pipeline, &cur_state, NULL, GST_CLOCK_TIME_NONE);
+  EXPECT_EQ (ret, GST_STATE_CHANGE_SUCCESS);
+  gst_object_unref (pipeline);
+
+  g_free (topic_name);
+  g_free (str_pipeline);
+  g_free (caps_str);
+}
+
+/**
+ * @brief Fail test case for mqttsrc #2 (MQTTAsync_unsubscribe failure case)
+ */
+TEST (testMqttSrc, src_normal_launch2)
+{
+  const gsize len_buf = 1024;
+  gchar *caps_str = g_strdup ("video/x-raw,width=640,height=320,format=RGB");
+  gchar *topic_name = g_strdup ("test_topic");
+  gchar *str_pipeline = g_strdup_printf (
+      "mqttsrc sub-topic=%s debug=true is-live=true num-buffers=%d ! "
+      "capsfilter caps=%s ! videoconvert ! videoscale ! fakevideosink",
+      topic_name, 1, caps_str);
+  GError *err = NULL;
+  //GstBuffer *buf;
+  GstElement *pipeline;
+  GstStateChangeReturn ret;
+  GstState cur_state;
+  GstMQTTMessageHdr hdr;
+  MQTTAsync_message *msg;
+  std::future<int> ma_ret;
+
+  pipeline = gst_parse_launch (str_pipeline, &err);
+  ASSERT_FALSE (pipeline == NULL);
+  ASSERT_TRUE (err == NULL);
+
+  GstMqttTestHelper::getInstance ().initFailFlags ();
+  GstMqttTestHelper::getInstance ().setFailUnsubscribe (TRUE);
+
+  msg = (MQTTAsync_message *) g_try_malloc0 (sizeof(*msg));
+  ASSERT_FALSE (msg == NULL);
+
+  _set_ts_gst_mqtt_message_hdr (pipeline, &hdr, GST_SECOND, 500 * GST_MSECOND);
+  ret = gst_element_set_state (pipeline, GST_STATE_PAUSED);
+  EXPECT_NE (ret, GST_STATE_CHANGE_FAILURE);
+
+  ret = gst_element_get_state (pipeline, &cur_state, NULL, GST_CLOCK_TIME_NONE);
+  EXPECT_EQ (ret, GST_STATE_CHANGE_NO_PREROLL);
+  EXPECT_EQ (cur_state, GST_STATE_PAUSED);
+
+  memset (hdr.gst_caps_str, '\0', GST_MQTT_MAX_LEN_GST_CPAS_STR);
+  memcpy (hdr.gst_caps_str, caps_str,
+      MIN (strlen (caps_str), GST_MQTT_MAX_LEN_GST_CPAS_STR - 1));
+  hdr.num_mems = 1;
+  hdr.size_mems[0] = len_buf;
+
+  msg->payloadlen = GST_MQTT_LEN_MSG_HDR + len_buf;
+  msg->payload = (MQTTAsync_message *) g_try_malloc0 (msg->payloadlen);
+  ASSERT_FALSE (msg->payload == NULL);
+
+  ret = gst_element_set_state (pipeline, GST_STATE_PLAYING);
+  EXPECT_NE (ret, GST_STATE_CHANGE_FAILURE);
+
+  {
+    gboolean mapped;
+    GstBuffer *buf;
+    GstMemory *mem;
+    GstMapInfo map;
+
+    buf = gst_buffer_new_allocate (NULL, len_buf, NULL);
+    ASSERT_FALSE (buf == NULL);
+
+    mem = gst_buffer_get_all_memory (buf);
+    ASSERT_FALSE (mem == NULL);
+
+    mapped = gst_memory_map (mem, &map, GST_MAP_READ);
+    ASSERT_EQ (mapped, TRUE);
+
+    memcpy (msg->payload, &hdr, GST_MQTT_LEN_MSG_HDR);
+    memcpy (&((guint8 *) msg->payload)[GST_MQTT_LEN_MSG_HDR], map.data,
+      len_buf);
+
+    gst_memory_unmap (mem, &map);
+    gst_buffer_unref (buf);
+  }
```

Comment:
This block seems to be redundant. How about wrapping this in any static function?

Reply:
@dongju-chae Right. I will use a static function wapping this block.